Update isIssuedBy and isSignedBy to use util::StatusOr

Martin2112 opened the pull request:
And corresponding change to isSelfSigned + update affected tests.

pphaneuf commented on cpp/log/cert.cc:
```diff
@@ -402,14 +401,14 @@ Cert::Status Cert::IsSignedBy(const Cert& issuer) const {
       LOG(WARNING) << "Unsupported algorithm: " << PrintSignatureAlgorithm();
       ClearOpenSSLErrors();
-      return UNSUPPORTED_ALGORITHM;
+      return util::Status(Code::UNAVAILABLE, "Unsupported algorithm");
```
Remember to make it `UNIMPLEMENTED` (and the other few spots).

pphaneuf commented on cpp/log/cert_checker.cc:
```diff
@@ -306,26 +307,27 @@ Status CertChecker::GetTrustedCa(CertChain* chain) const {
-    Cert::Status ok = subject->IsSignedBy(*issuer_cand);
-    if (ok == Cert::UNSUPPORTED_ALGORITHM) {
+    StatusOr<bool> signed_by_issuer = subject->IsSignedBy(*issuer_cand);
+    if (signed_by_issuer.status().error_code() == Code::UNAVAILABLE) {
```
Use `CanonicalCode()` instead of `error_code()`?

pphaneuf commented on cpp/log/cert_test.cc:
```diff
@@ -351,7 +352,7 @@ TEST_F(CertTest, TestUnsupportedAlgorithm) {
 // user that their setup is insecure.
 #ifdef OPENSSL_NO_MD2
-  EXPECT_EQ(Cert::UNSUPPORTED_ALGORITHM, legacy.IsSignedBy(legacy));
+  EXPECT_EQ(Code::UNAVAILABLE, legacy.IsSignedBy(legacy).status().error_code());
```
For `util::Status` tests, you can use `EXPECT_THAT`, with the `StatusIs` helper (from `cpp/util/status_test_util.h`, something like this:

```C++
EXPECT_THAT(legacy.IsSignedBy(legacy).status(), StatusIs(Code::UNAVAILABLE));
```

Yeah, they flip the expected/actual with `EXPECT_THAT`. Gotta keep it exciting! :smiley:

pphaneuf commented:
All reviewed, mostly nits, really.

Martin2112 commented:
I think I've fixed everything. Also rebased and fixed up conflicts with other refactorings. It's starting to pay off a bit e.g look at CertChain::IsValidSignatureChain.

pphaneuf commented on cpp/log/cert_checker.cc:
```diff
@@ -292,26 +293,27 @@ Status CertChecker::GetTrustedCa(CertChain* chain) const {
-    Cert::Status ok = subject->IsSignedBy(*issuer_cand);
-    if (ok == Cert::UNSUPPORTED_ALGORITHM) {
+    StatusOr<bool> signed_by_issuer = subject->IsSignedBy(*issuer_cand);
+    if (signed_by_issuer.status().CanonicalCode() == Code::UNIMPLEMENTED) {
```
I wonder if we shouldn't just pass back the underlying status, or if in fact, this sort of translation is good (to control better what the end user see in the HTTP responses)?

Martin2112 replied:
It's not an easy one to answer. In some cases the low level error can be
translated to something more meaningful at a higher level. OTOH you can end
up losing the actual cause of the problem by returning something generic.

On 4 September 2015 at 11:37, Pierre Phaneuf <notifications@github.com>
wrote:

> In cpp/log/cert_checker.cc
> <https://github.com/google/certificate-transparency/pull/905#discussion_r38739808>
> :
>
> > @@ -292,26 +293,27 @@ Status CertChecker::GetTrustedCa(CertChain* chain) const {
> >         it != issuer_range.second; ++it) {
> >      const Cert* issuer_cand = it->second;
> >
> > -    Cert::Status ok = subject->IsSignedBy(*issuer_cand);
> > -    if (ok == Cert::UNSUPPORTED_ALGORITHM) {
> > +    StatusOr<bool> signed_by_issuer = subject->IsSignedBy(*issuer_cand);
> > +    if (signed_by_issuer.status().CanonicalCode() == Code::UNIMPLEMENTED) {
>
> I wonder if we shouldn't just pass back the underlying status, or if in
> fact, this sort of translation is good (to control better what the end user
> see in the HTTP responses)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/certificate-transparency/pull/905/files#r38739808>
> .
>

pphaneuf commented:
LGTM
